### PR TITLE
more buns

### PR DESCRIPTION
**Pull request by ioga**

### CODE REVIEW

**Comment by ioga** [comment]
1. _make it more “bun”_. I think “your” code with bun tries to mimic what we do in master today - a centralized db object that holds everything - but with bun. It has got its own DB objects, its own RunInTx. I don’t think it adds this much, and it eats `context` which we probably need to start using. I suggest to instead write modular code which is more like [bun’s own examples](https://github.com/uptrace/bun/blob/master/example/tx-composition/main.go). I’d like us to write “bun” database code, not “our” database code.
2. _make our db singleton more like context.background_. Singletony `db.DB` (which is `bun.DB` instance)  allows you to use db everywhere (just like context.Background() is your bailout for context)
3. _optional `Tx` suffix when you need complexity_. if you want a method to be used in transactions or cancellable contexts, have a `Tx` suffixed version. if you want to ignore context and tx, you can use straight up db.DB & ctx.Background. like this: https://github.com/determined-ai/ci-shortcuts/pull/2/files#diff-3197170506f9f7dee1a371a87f8abda06d7122b8c9a66b5c89e78dd9d7e3aa09R45 and its unsuffixed thin wrapper: https://github.com/determined-ai/ci-shortcuts/pull/2/files#diff-3197170506f9f7dee1a371a87f8abda06d7122b8c9a66b5c89e78dd9d7e3aa09R55
4. I’ve exaggerated modularity by putting artifacts and builds into separate subpackages.
5. A lot of methods in this code ended up being “static” top-level functions - but there’re “object” methods as well, e.g. https://github.com/determined-ai/ci-shortcuts/pull/2/files#diff-3197170506f9f7dee1a371a87f8abda06d7122b8c9a66b5c89e78dd9d7e3aa09R100

File: circleci.go [hunk 1]
```diff
@@ -148,15 +113,15 @@ func (c *Client) BootstrapBuilds(db RootDB) error {
 					fmt.Printf(
 						"Quitting after 30 days of builds; artifacts won't exist further back.\n",
 					)
-					goto report
+					goto report // WTF
```
Comment (stoksc) [inline circleci.go]: lol

**Review by stoksc — COMMENTED** [review]
lgtm. i like the singleton a lot, actually.

**Comment by stoksc** [comment]
style wise, it's super interesting. not sure how i feel yet.

File: db/db.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	_ "github.com/mattn/go-sqlite3"
+
+	"github.com/uptrace/bun"
+	"github.com/uptrace/bun/dialect/sqlitedialect"
+	"github.com/uptrace/bun/extra/bundebug"
+)
+
+var DB *RootDB
+
+type RootDB = bun.DB
```
Comment (rb-determined-ai) [inline db/db.go]: why bother with this type alias?

File: db/db.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	if err != nil {
+		sqldb.Close()
+		return RootDB{}, errors.Wrap(err, "creating tables")
+	}
+
+	db := bun.NewDB(sqldb, sqlitedialect.New())
+	// db.AddQueryHook(bundebug.NewQueryHook(bundebug.WithVerbose(true)))
+	db.AddQueryHook(bundebug.NewQueryHook())
+	DB = db
+	return *DB, nil
```
Comment (rb-determined-ai) [inline db/db.go]: `*DB` strikes me as "not safe in general"... do you know that passing bun.DB by value is actually acceptable?

File: db/db.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+type RootDB = bun.DB
+
+func NewDB(dbPath string) (RootDB, error) {
+	sqldb, err := sql.Open("sqlite3", dbPath)
+	if err != nil {
+		return RootDB{}, err
+	}
+
+	// create the database if it doesn't exist
+	// XXX this'll be in bun migrations.
```
Comment (rb-determined-ai) [inline db/db.go]: It sounds like we're fine using bun migrations IRL, but I have zero intention of allowing migrations on this webserver.

Blowing away the sqlite.db and starting from scratch results in degraded response times for like 30 min, and I expect that to happen maybe once a year.  Not worth the complexity.

Comment (ioga) [reply db/db.go]: yep I added this comment mostly for illustration db.go won't have to contain a description of all models in the database.

File: circleci.go (hunk 1 above)
Comment (rb-determined-ai) [reply circleci.go]: cleanest way to break out of two loops.  Some modern languages have labeled loops so you can `break outer_loop;` but go is too cool for that.

File: db/db.go (hunk 2 above)
Comment (ioga) [reply db/db.go]: this is an omission - I didn't get rid of RootDB and just aliased it so don't need to refactor it.

File: db/db.go (hunk 3 above)
Comment (ioga) [reply db/db.go]: this is also an omission - I did it like this just to make it compatible with the other code.

**Comment by rb-determined-ai** [comment]
I like this in general.  One thing I noticed was that most of the functions I was writing in the BunDB object in master were extremely simple (that is, pointless).  Meanwhile I wasn't even able to cleanly separate the bun code within the BunDB object... I was leaking Bun-like details to achieve the flexibility I wanted.

If you like, I can land my pr and you can land this on top of it.  But there's some `// WTF`s and `// XXX`s that ought to be removed.

**Pull request merged**